x/staking/types/errors.go

```diff
@@ -11,44 +11,62 @@ import (
+	ErrBadDelegatorAddr                         = sdkerrors.Register(ModuleName, 20, "delegator does not exist with address")
+	ErrNoDelegatorForAddress                    = sdkerrors.Register(ModuleName, 21, "delegator does not contain delegation")
+	ErrInsufficientShares                       = sdkerrors.Register(ModuleName, 22, "insufficient delegation shares")
+	ErrDelegationValidatorEmpty                 = sdkerrors.Register(ModuleName, 23, "cannot delegate to an empty validator")
+	ErrNotEnoughDelegationShares                = sdkerrors.Register(ModuleName, 24, "not enough delegation shares")
+	ErrNotMature                                = sdkerrors.Register(ModuleName, 25, "entry not mature")
+	ErrNoUnbondingDelegation                    = sdkerrors.Register(ModuleName, 26, "no unbonding delegation found")
+	ErrMaxUnbondingDelegationEntries            = sdkerrors.Register(ModuleName, 27, "too many unbonding delegation entries for (delegator, validator) tuple")
+	ErrNoRedelegation                           = sdkerrors.Register(ModuleName, 28, "no redelegation found")
+	ErrSelfRedelegation                         = sdkerrors.Register(ModuleName, 29, "cannot redelegate to the same validator")
+	ErrTinyRedelegationAmount                   = sdkerrors.Register(ModuleName, 30, "too few tokens to redelegate (truncates to zero tokens)")
+	ErrBadRedelegationDst                       = sdkerrors.Register(ModuleName, 31, "redelegation destination validator not found")
+	ErrTransitiveRedelegation                   = sdkerrors.Register(ModuleName, 32, "redelegation to this validator already in progress; first redelegation to this validator must complete before next redelegation")
+	ErrMaxRedelegationEntries                   = sdkerrors.Register(ModuleName, 33, "too many redelegation entries for (delegator, src-validator, dst-validator) tuple")
+	ErrDelegatorShareExRateInvalid              = sdkerrors.Register(ModuleName, 34, "cannot delegate to validators with invalid (zero) ex-rate")
+	ErrBothShareMsgsGiven                       = sdkerrors.Register(ModuleName, 35, "both shares amount and shares percent provided")
+	ErrNeitherShareMsgsGiven                    = sdkerrors.Register(ModuleName, 36, "neither shares amount nor shares percent provided")
+	ErrInvalidHistoricalInfo                    = sdkerrors.Register(ModuleName, 37, "invalid historical info")
+	ErrNoHistoricalInfo                         = sdkerrors.Register(ModuleName, 38, "no historical info found")
+	ErrEmptyValidatorPubKey                     = sdkerrors.Register(ModuleName, 39, "empty validator public key")
+	ErrUnbondingNotFound                        = sdkerrors.Register(ModuleName, 40, "unbonding operation not found")
+	ErrUnbondingOnHoldRefCountNegative          = sdkerrors.Register(ModuleName, 41, "cannot un-hold unbonding operation that is not on hold")
+	ErrNotEnoughBalance                         = sdkerrors.Register(ModuleName, 42, "not enough balance")
```

Comment:
These error codes already colliding with what is done for 0.46, 0.47 and planned for 0.50, see:
https://github.com/cosmos/cosmos-sdk/blob/dd585bac075596cc075643e36366e4ed2381c6c9/x/staking/types/errors.go#L47-L50

I suggest setting LSM-related errors into own prefix (`1xx`):

```go
	ErrNotEnoughBalance                         = sdkerrors.Register(ModuleName, 101, "not enough balance")
	ErrTokenizeShareRecordNotExists             = sdkerrors.Register(ModuleName, 102, "tokenize share record not exists")
	ErrTokenizeShareRecordAlreadyExists         = sdkerrors.Register(ModuleName, 103, "tokenize share record already exists")
	ErrNotTokenizeShareRecordOwner              = sdkerrors.Register(ModuleName, 104, "not tokenize share record owner")
	ErrExceedingFreeVestingDelegations          = sdkerrors.Register(ModuleName, 105, "trying to exceed vested free delegation for vesting account")
	ErrOnlyBondDenomAllowdForTokenize           = sdkerrors.Register(ModuleName, 106, "only bond denom is allowed for tokenize")
	ErrInsufficientValidatorBondShares          = sdkerrors.Register(ModuleName, 107, "insufficient validator bond shares")
	ErrRedelegationNotAllowedForValidatorBond   = sdkerrors.Register(ModuleName, 108, "redelegation is not allowed for validator bond delegation")
	ErrValidatorBondNotAllowedForTokenizeShare  = sdkerrors.Register(ModuleName, 109, "validator bond delegation is not allowed to tokenize share")
	ErrValidatorBondNotAllowedFromModuleAccount = sdkerrors.Register(ModuleName, 110, "validator bond is not allowed from a module account")
	ErrGlobalLiquidStakingCapExceeded           = sdkerrors.Register(ModuleName, 111, "delegation from liquid staking provider exceeds the global cap")
	ErrValidatorLiquidStakingCapExceeded        = sdkerrors.Register(ModuleName, 112, "delegation from liquid staking provider exceeds the validator cap")
	ErrTokenizeSharesDisabledForAccount         = sdkerrors.Register(ModuleName, 113, "tokenize shares currently disabled for account")
	ErrUnableToDisableTokenizeShares            = sdkerrors.Register(ModuleName, 114, "unable to disable tokenize shares for account")
	ErrTokenizeSharesAlreadyEnabledForAccount   = sdkerrors.Register(ModuleName, 115, "tokenize shares is already enabled for this account")
	ErrTokenizeSharesAlreadyDisabledForAccount  = sdkerrors.Register(ModuleName, 116, "tokenize shares is already disabled for this account")
	ErrValidatorLiquidSharesUnderflow           = sdkerrors.Register(ModuleName, 117, "validator liquid shares underflow")
	ErrTotalLiquidStakedUnderflow               = sdkerrors.Register(ModuleName, 118, "total liquid staked underflow")
```

Reply:
updated to your suggests [here](https://github.com/cosmos/cosmos-sdk/pull/16747/commits/7b4eac53bae3069f0db6de4fbf0df64354fb9711)